### PR TITLE
numpydoc style(signac/contrib/project.py)

### PR DESCRIPTION
**Pull request by ac-optimus**
## Description

issue #315 
## Types of Changes

- [x] Documentation update
- [ ] Bug fix
- [ ] New feature
- [ ] Breaking change<sup>1</sup>

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:

- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac/blob/master/contributors.yaml).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md#code-style) of this project.
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [ ] I have updated the API documentation as part of the package doc-strings.
- [ ] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [ ] I have updated the [changelog](https://github.com/glotzerlab/signac/blob/master/changelog.txt) and added all related issue and pull request numbers for future reference (if applicable). See example below.

Change of doc string to `numpydoc` in schema.py

### CODE REVIEW

**Comment by csadorf** [comment]
@ac-optimus I see that there are a lot of *empty* doc-strings. I assume that you need guidance with those? What is the proposed workflow to complete those?

**Comment by ac-optimus** [comment]
> @ac-optimus I see that there are a lot of _empty_ doc-strings. I assume that you need guidance with those? What is the proposed workflow to complete those?

hi, as per the discussion with @vyasr on slack -- he asked to fill all the doc strings that I understand and wait for someone more experienced in the file to get  the feedback on the doc-strings that I am less sure about.

File: signac/contrib/project.py [hunk 1]
```diff
@@ -629,25 +896,42 @@ def find_jobs(self, filter=None, doc_filter=None):
...
+            A mapping of key-value pairs that all
+            indexed job statepoints are compared against. (Default value = None)
+        doc_filter : Mapping
+            A mapping of key-value pairs that all
+            indexed job documents are compared against. (Default value = None)
+
+        Yields
+        ------
+        Instances of :class:`~signac.contrib.job.Job`
+
```
Comment (bdice) [inline signac/contrib/project.py:911]: This should be removed in favor of the `Returns` section. This isn't a generator itself, but it returns one.
```suggestion
```

File: signac/contrib/project.py [hunk 2]
```diff
@@ -629,25 +896,42 @@ def find_jobs(self, filter=None, doc_filter=None):
...
+        Returns
+        -------
+            JobsCursor ka eek tyep return hota ahi
```
Comment (bdice) [inline signac/contrib/project.py:914]:
```suggestion
            JobsCursor of jobs matching the provided filter(s).
```

File: signac/contrib/project.py [hunk 3]
```diff
@@ -629,25 +896,42 @@ def find_jobs(self, filter=None, doc_filter=None):
...
+
         """
         return JobsCursor(self, filter, doc_filter)
 
     def __iter__(self):
         return iter(self.find_jobs())
 
     def groupby(self, key=None, default=None):
-        """Groups jobs according to one or more statepoint parameters.
+        """Group jobs according to one or more statepoint parameters.
```
Comment (bdice) [inline signac/contrib/project.py:933]:
```suggestion
        """Group jobs according to one or more state point parameters.
```

File: signac/contrib/project.py [hunk 4]
```diff
@@ -672,19 +956,25 @@ def groupby(self, key=None, default=None):
         If `key` is None, jobs are grouped by identity (by id), placing one job
         into each group.
 
-        :param key:
+        Parameters
+        ----------
+        key : str, iterable, or function
             The statepoint grouping parameter(s) passed as a string, iterable of strings,
```
Comment (bdice) [inline signac/contrib/project.py:962]:
```suggestion
            The state point grouping parameter(s) passed as a string, iterable of strings,
```

File: signac/contrib/project.py [hunk 5]
```diff
@@ -707,32 +997,54 @@ def groupbydoc(self, key=None, default=None):
...
         return self.find_jobs().groupbydoc(key, default=default)
 
     def to_dataframe(self, *args, **kwargs):
         """Export the project metadata to a pandas dataframe.
 
         The arguments to this function are forwarded to :py:meth:`.JobsCursor.to_dataframe`.
+
+        Parameters
+        ----------
+        *args :
```
Comment (bdice) [inline signac/contrib/project.py:1020]: I don't know how this appears when formatted in the documentation. We'll need to review all the docs when rendered with Sphinx, but I don't want to solve this right now.

File: signac/contrib/project.py [hunk 6]
```diff
@@ -707,32 +997,54 @@ def groupbydoc(self, key=None, default=None):
...
+
+        **kwargs :
+
+
+        Returns
+        -------
+            class 'pandas.core.frame.DataFrame'
```
Comment (bdice) [inline signac/contrib/project.py:1027]:
```suggestion
            pandas.DataFrame
```

File: signac/contrib/project.py [hunk 7]
```diff
@@ -707,32 +997,54 @@ def groupbydoc(self, key=None, default=None):
...
+
         """
         return self.find_jobs().to_dataframe(*args, **kwargs)
 
     def read_statepoints(self, fn=None):
         """Read all statepoints from a file.
 
-        :param fn: The filename of the file containing the statepoints,
+        See also :meth:`dump_statepoints` and :meth:`write_statepoints`.
```
Comment (bdice) [inline signac/contrib/project.py:1035]: Use the "See Also" section mentioned here: https://numpydoc.readthedocs.io/en/latest/format.html#sections

File: signac/contrib/project.py [hunk 8]
```diff
@@ -707,32 +997,54 @@ def groupbydoc(self, key=None, default=None):
...
+        fn : str
+            The filename of the file containing the statepoints,
             defaults to :const:`~signac.contrib.project.Project.FN_STATEPOINTS`.
-        :type fn: str
 
-        See also :meth:`dump_statepoints` and :meth:`write_statepoints`.
+        Returns
+        -------
+        dict
+            state ponit.
```
Comment (bdice) [inline signac/contrib/project.py:1046]:
```suggestion
            State points.
```

File: signac/contrib/project.py [hunk 9]
```diff
@@ -762,17 +1080,19 @@ def write_statepoints(self, statepoints=None, fn=None, indent=2):
 
         If the file already contains statepoints, all new statepoints
         will be appended, while the old ones are preserved.
+        See also :meth:`dump_statepoints`.
```
Comment (bdice) [inline signac/contrib/project.py:1083]: Use the "See Also" section: https://numpydoc.readthedocs.io/en/latest/format.html#sections

**Comment by ac-optimus** [comment]
> @ac-optimus Here's a partial review. I got about halfway through the PR, but it's very large. After you apply these changes, please make another pass over the full PR. See if you can apply the same patterns from my suggested changes for other parts of the `project.py` file that I didn't finish reviewing. Then re-request review. @jennyfothergill If you can provide some intermediate feedback on this PR, please start from the _bottom_ of `project.py`. I posted some guidelines for reviewing numpydoc pull requests in the signac slack: https://signac.slack.com/archives/CPEHE14R5/p1586373911096900

yes sure!

File: signac/contrib/project.py [hunk 10]
```diff
@@ -518,7 +720,21 @@ def build_job_statepoint_index(self, exclude_const=False, index=None):
 
         .. code-block:: python
 
-            >>> for i in range(4):
+
+        Values that are constant over the complete data space can be optionally
+        ignored with the exclude_const argument set to True.
+
+        Parameters
```
[the remark this answers is not in the document]
Comment (ac-optimus) [reply signac/contrib/project.py:731]: made the changes to other files accordingly.

File: signac/contrib/project.py [hunk 11]
```diff
@@ -629,29 +898,44 @@ def find_jobs(self, filter=None, doc_filter=None):
...
+            compared against (Default value = None).
+
+        Returns
+        -------
+        signac.contrib.project.JobsCursor
+            JobsCursor of jobs matching the provided filter(s).
+
+        Raises
+        ------
+        TypeError:
```
Comment (jennyfothergill) [inline signac/contrib/project.py:917]: This is super nitpicky, but sometimes the errors have a colon after them and sometimes they don't. E.g., line 917 is `TypeError:` but line 873 is `TypeError`

File: signac/contrib/project.py [hunk 12]
```diff
@@ -87,6 +108,23 @@ def _resolve_statepoint_filter(self, q):
                 yield 'statepoint.{}'.format(k), v
 
     def find_job_ids(self, filter=None, doc_filter=None):
+        """Find job ids from a state point or document filter.
+
+        Parameters
+        ----------
+        filter : dict
+            A mapping of key-value pairs that all indexed job statepoints are
```
Comment (jennyfothergill) [inline signac/contrib/project.py:116]: There's some inconsistency with statepoint and state point. Do they mean different things?

Comment (jennyfothergill) [reply signac/contrib/project.py:116]: line 1039 is "state point", but here is "statepoint"

Comment (bdice) [reply signac/contrib/project.py:116]: @jennyfothergill We generally try to use "state point" as two words, but this is not consistent across the docs. The property is named "statepoint", which is a source of confusion.

Comment (bdice) [reply signac/contrib/project.py:116]:
```suggestion
            A mapping of key-value pairs that all indexed job state points are
```

File: signac/contrib/project.py (hunk 11 above)
Comment (bdice) [reply signac/contrib/project.py:917]: I think no colon is correct, based on the style guide here: https://numpydoc.readthedocs.io/en/latest/format.html
```suggestion
        TypeError
```

Comment (vyasr) [reply signac/contrib/project.py:917]: I agree, I think I removed this elsewhere.

Comment (ac-optimus) [reply signac/contrib/project.py]: corrected it.

File: signac/contrib/project.py (hunk 12 above)
Comment (ac-optimus) [reply signac/contrib/project.py]: changed `"statepoint"/ "statepoints"` to `"state point"/ "state points"` at all the required places.

**Comment by ac-optimus** [comment]
> Wow, this is a really big PR--nice work @ac-optimus! I'm not familiar enough with the code to verify that the docstrings are correct, but I noticed a couple minor inconsistencies in the format. This PR will definitely make the documentation more accessible. :)

Thanks! Also, I have addressed all the inconsistencies :)

File: signac/contrib/project.py [hunk 13]
```diff
@@ -65,18 +67,37 @@ class JobSearchIndex(object):
...
 
     def _resolve_statepoint_filter(self, q):
+        """Resolve state point based on filter.
+
+        Parameters
+        ----------
+        q : dict
+            Filter used for state point selection.
+
+        Yield
```
Comment (vyasr) [inline signac/contrib/project.py:95]: Unfortunately I think this has to be plural
```suggestion
        Yields
```

File: signac/contrib/project.py [hunk 14]
```diff
@@ -200,28 +257,57 @@ def __eq__(self, other):
...
+        """
         return self._config['project_dir']
 
     @property
     def _wd(self):
+        """Get Project workspace directory.
+
+        Returns
+        -------
+        str :
```
Comment (vyasr) [inline signac/contrib/project.py:288]:
```suggestion
        str
```

File: signac/contrib/project.py [hunk 15]
```diff
@@ -200,28 +257,57 @@ def __eq__(self, other):
...
         else:
             return os.path.join(self._rd, wd)
 
     def root_directory(self):
-        "Returns the project's root directory."
+        """Return the project's root directory.
+
+        Returns
+        -------
+        str :
```
Comment (vyasr) [inline signac/contrib/project.py:303]:
```suggestion
        str
```

File: signac/contrib/project.py [hunk 16]
```diff
@@ -65,18 +67,37 @@ class JobSearchIndex(object):
...
     def _resolve_statepoint_filter(self, q):
+        """Resolve state point based on filter.
+
+        Parameters
+        ----------
+        q : dict
+            Filter used for state point selection.
+
+        Yields
+        -----
```
Comment (vyasr) [inline signac/contrib/project.py:96]:
```suggestion
        ------
```

**Pull request merged**